python/lsst/daf/butler/queries/tree/_query_tree.py

```
@@ -0,0 +1,314 @@
+        data_coordinate_uploads = dict(self.data_coordinate_uploads)
+        data_coordinate_uploads[key] = dimensions
+        return self.model_copy(
+            update=dict(
+                dimensions=self.dimensions | dimensions, data_coordinate_uploads=data_coordinate_uploads
+            )
+        )
+
+    def join_materialization(self, key: MaterializationKey, dimensions: DimensionGroup) -> QueryTree:
+        """Return a new tree that joins in temporarily stored results from
+        another query.
+
+        Parameters
+        ----------
+        key : `MaterializationKey`
+            Unique identifier for this materialization, as assigned by a
+            `QueryDriver`.
+        dimensions : `DimensionGroup`
+            The dimensions stored in the materialization.
+
+        Returns
+        -------
+        result : `QueryTree`
+            A new tree that joins in the materialization.
+        """
+        assert key not in self.data_coordinate_uploads, "Query should prevent duplicate materialization."
+        materializations = dict(self.materializations)
+        materializations[key] = dimensions
+        return self.model_copy(
+            update=dict(dimensions=self.dimensions | dimensions, materializations=materializations)
+        )
+
+    def join_dataset(self, dataset_type: str, search: DatasetSearch) -> QueryTree:
+        """Return a new tree joins in a search for a dataset.
+
+        Parameters
+        ----------
+        dataset_type : `str`
+            Name of dataset type to join in.
+        search : `DatasetSearch`
+            Struct containing the collection search path and dataset type
+            dimensions.
+
+        Returns
+        -------
+        result : `QueryTree`
+            A new tree that joins in the dataset search.
+
+        Notes
+        -----
+        If this dataset type was already joined in, the new `DatasetSearch`
+        replaces the old one.
```

I wonder if it should raise an exception instead?  Presumably the existing DatasetSearch was added for reasons, so silently discarding it could lead to potentially confusing bugs.  I see that you guard against this at the higher level but as the code churns it might be good to check it here too.

---

I'll make it assert.  It used to raise an exception with a friendly error message then I dropped that to make it as simple as possible when trying to remove duplication between `QueryTree` and `Query`; I probably just went a bit too far in this case.

---

Done on DM-43146.